### PR TITLE
Patch 14734 boolean sync validators

### PR DESCRIPTION
**Pull request by BobChao87**
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**

Fix/clarification of how synchronous validators work.

**What is the current behavior? (You can also link to an open issue here)**
#14734 All values except `false` literal and `undefined` are treated as a pass. `undefined` is treated as pending, which makes no sense for synchronous operation.

**What is the new behavior (if this is a feature change)?**

All falsy value returns to a synchronous validator are treated as fails, including `undefined`.

**Does this PR introduce a breaking change?**

Yes, see commit message.

**Please check if the PR fulfills these requirements**
- [ x ] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [ x ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Other information**:

I believe a change to docs in unnecessary. We still expect `true` and `false` to be the primary returns and don't want to encourage otherwise. This is mostly clarifying a niche, undocumented, edge-case to behave more intuitively.

### CODE REVIEW

File: test/ng/directive/ngModelSpec.js [hunk 1]
```diff
@@ -847,6 +847,38 @@ describe('ngModel', function() {
         expect(ctrl.$valid).toBe(true);
       });
 
+      it('should treat all responses as boolean for synchronous validators', function() {
+        var curry = function(v) {
```
Comment (gkalpak) [inline test/ng/directive/ngModelSpec.js]: You don't need a new function for this. You can use the internal helper: `valueFn()`

File: test/ng/directive/ngModelSpec.js [hunk 2]
```diff
@@ -847,6 +847,38 @@ describe('ngModel', function() {
...
+          return function() {
+            return v;
+          };
+        };
+
+        var tester = function(v, e) {
```
Comment (gkalpak) [inline test/ng/directive/ngModelSpec.js]: Nit: I would prefer a more descriptive name, such as `expectValid()`.

**Review by gkalpak — COMMENTED** [review]
Apart from a couple of nitpicks, LGTM.

<sub>I would add a little more details to the commit message (regarding what was happening before) - can be done while merging.</sub>

File: test/ng/directive/ngModelSpec.js (hunk 1 above)
Comment (BobChao87) [inline test/ng/directive/ngModelSpec.js]: Interesting, I'll make the switch. I was actually hoping to talk about the `curry()` anyways, because it's now used three times in this file (hence why I used it in the first place). I'll submit a separate PR to fix the others.

Comment (gkalpak) [inline test/ng/directive/ngModelSpec.js]: Good catch! Feel free to change all other instances too.

**Comment by BobChao87** [comment]
Hopefully those changes to the commit message are in the direction you wanted.

**Review by gkalpak — APPROVED** [review]
LGTM

**Comment by petebacondarwin** [comment]
Landed thanks.
